README.md

```diff
@@ -157,7 +157,7 @@ We export `COLUMNS` because `se lint` needs to know the width of the terminal so
 Before we can use `pylint` or `mypy` on the toolset source, we have to inject them into the venv `pipx` created for the `standardebooks` package:
 
 ```shell
-pipx inject standardebooks pylint mypy
+pipx inject standardebooks pylint mypy types-setuptools types-requests
```

Comment:
I'm not super happy about having to inject some mystery packages into the venv just to get mypy working. Is there any way around this?

The more important question is, why did the tests start failing even though we didn't change anything? What changed and how can we prevent changes from happening by surprise in the future? The testing infrastructure should be pinned to a single version of python and friends, which should not shift underfoot randomly.

Reply:
The tests started failing because the Ubuntu image stopped being able to find Python 3.9.5. This first started in https://github.com/standardebooks/tools/actions/runs/1034760856. I don’t know what caused that failure, but 3.9.6 was marked as available.

While I was upgrading that anyway, I thought it’d be good to bump the test package versions to the latest. We don’t specific versions in the README, so I assume that local testing will pull in whatever the latest is. We can leave mypy pinned at the current 0.812, but any upgrade from that will hit the problem of the extra packages. More info in https://mypy-lang.blogspot.com/2021/05/the-upcoming-switch-to-modular-typeshed.html.

I’m of course happy to downgrade mypy to 0.812, but we’d also want to fix that version in the README instructions so that there aren’t major test differences between local and github.

Reply:
Does that mean that the testing framework was installing a new version of Ubuntu? If so then we need to pin the Ubuntu version to 20.04 which is the current LTS. Is there a way to do that? Bottom line, the testing framework should not get a surprise version of Python every time it spins up and AFAIK the only way that can happen is across distro releases.

Reply:
No, the Ubuntu image is already pinned to 20.04: https://github.com/standardebooks/tools/blob/fed7af0e29a5fdb42dd0c77fb022f9c32f3c3323/.github/workflows/se-test.yml#L8. I honestly have no idea why 3.9.5 just disappeared from Github’s Ubuntu image. I’ve gone looking, but can’t find anyone talking about it.

Reply:
Looking at that config file, it looks like Python is installed at the Github Actions level and not on the distro level? `with: .. python-version` applies to Github Actions, that's not an Ubuntu level command. So perhaps the question is, what is actually going on in Github Actions when it spins up an image and how do we *really* pin Python versions there?

Reply:
The missing version seems to be in the current manifest, and the file exists: https://github.com/actions/python-versions/blob/e68c876eefde5bde961e4a1d8090b90f276c78e3/versions-manifest.json#L572-L578

As far as I’m concerned, the version pinning should just work. But this isn’t unfortunately my core competence.

Reply:
Posted an issue to github.community to see if anyone else has run into this: https://github.community/t/missing-python-version-3-9-5/190838

Reply:
Looking at https://github.com/actions/setup-python it looks like that step is only for installing a different version of Python. What if we remove it entirely? Presumably the OS default Python will be used instead? In any case, let's move this into a separate PR because we need to fix this in a way where we can actually pin a Python version permanently.